### PR TITLE
Implement UpperHex, LowerHex, Octal and Binary

### PR DESCRIPTION
**Pull request by KaiWitt**
Issue #87 
Implementation of `fmt::{UpperHex, LowerHex, Octal, Binary} ` for each num type.

### CODE REVIEW

File: num/src/bigint.rs [hunk 1]
```diff
@@ -848,6 +848,54 @@ macro_rules! construct_bigint {
             }
         }
 
+        impl core::fmt::UpperHex for $name {
+            fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+                if f.alternate() {
+                    write!(f, "0x")?;
+                }
+                for chunk in self.0.iter().rev() {
+                    write!(f, "{:0>16X}", chunk)?;
```
Comment (dr-orlovsky) [inline num/src/bigint.rs:857]: I think we need to follow formatting patters from `f`, i.e. the width and alignment instructions. This probably can be done with something like
`"{:0$widthX}", $value, width=f.width()`

File: num/src/bigint.rs [hunk 2]
```diff
@@ -848,6 +848,54 @@ macro_rules! construct_bigint {
             }
         }
 
+        impl core::fmt::UpperHex for $name {
+            fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+                if f.alternate() {
+                    write!(f, "0x")?;
```
Comment (dr-orlovsky) [inline num/src/bigint.rs:854]: If I am not wrong, rust uses `0X` prefix for upper hex - but I personally always prefer `0x`

**Review by dr-orlovsky — COMMENTED** [review]
Thank you very much! Left one suggestion, pls let me known what do you think

File: num/src/bigint.rs (hunk 2 above)
Comment (KaiWitt) [reply num/src/bigint.rs:854]: > If I am not wrong, rust uses `0X` prefix for upper hex

I think you're wrong on this one. The [documentation](https://doc.rust-lang.org/std/fmt/trait.UpperHex.html) uses `0x` and [code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f46f0c45d753f4c3a8fb04f8d0ee87de) aswell

File: num/src/bigint.rs (hunk 1 above)
Comment (KaiWitt) [reply num/src/bigint.rs:857]: Makes sense, I will add them

**Comment by KaiWitt** [comment]
The implementation now covers padding with the `0` flag. Other width padding and alignment instructions are delegated to `Formatter::pad`. I've also extended the test cases to cover widths and alignments

**Review by dr-orlovsky — APPROVED** [review]
Perfect, thank you very much!

**Comment by dr-orlovsky** [comment]
Sorry, we have CI fails in `no_std` env; so we need to import things like `alloc::string::String` either inside specific functions (preferred way) or at the top under `#[cfg(feature = "alloc")]` gates

**Comment by KaiWitt** [comment]
Ok, alloc String and format is now imported explicitly. Hope CI passes now

**Comment by dr-orlovsky** [comment]
Sorry for late merge: I am getting dozen of GitHub notifications each day and sometimes miss important stuff :( Pls feel free to ping if you see that some of your contributions has stalled

**Pull request merged**